Modules/posixmodule.c

```diff
@@ -5490,12 +5490,89 @@ os__path_normpath_impl(PyObject *module, PyObject *path)
         return NULL;
     }
     Py_ssize_t norm_len;
-    wchar_t *norm_path = _Py_normpath_and_size(buffer, len, &norm_len);
+    wchar_t *norm_path = _Py_normpath_and_size(buffer, len, 0, &norm_len);
     PyObject *result = PyUnicode_FromWideChar(norm_path, norm_len);
     PyMem_Free(buffer);
     return result;
 }
 
+#ifndef MS_WINDOWS
+
+/*[clinic input]
+os._path_abspath
+
+    path: unicode
+    /
+
+Make path absolute.
+[clinic start generated code]*/
+
+static PyObject *
+os__path_abspath_impl(PyObject *module, PyObject *path)
+/*[clinic end generated code: output=b58956d662b60be0 input=577ecb3473d22113]*/
+{
+    Py_ssize_t rel_path_len;
+    wchar_t *rel_path = PyUnicode_AsWideCharString(path, &rel_path_len);
+    if (!rel_path) {
+        return NULL;
+    }
+
+    if (rel_path[0] == '\0' || !wcscmp(rel_path, L".")) {
+        return posix_getcwd(0);
+    }
+
+    Py_ssize_t cwd_len;
+    wchar_t *abs_path;
+    Py_ssize_t abs_path_len;
+    if (_Py_isabs(rel_path)) {
+        cwd_len = 0;
+        abs_path = rel_path;
+        abs_path_len = rel_path_len;
+    }
```

Comment:
It's clearer to me if `rel_path` and `rel_path_len` are named `path_buf` and `path_buf_len`, because the input path may already be absolute and just needs to be normalized, or it may already be absolute and normalized.

Also, I'd prefer to add a `start` variable that gets set here instead of `cwd_len`. The block that actually uses the working directory would set `start = cwd_len`.